### ISSUE TITLE
Getting error randomly : function timed out, ensure the promise resolves within 180000 milliseconds

### ISSUE THREAD

**Issue by espekkaya** [angular/protractor#5172]
Hello,

We have got an error randomly for 2 weeks. I have not figured it out why. And the biggest problem is once it accrued in some point then the rest of suite got an error as well.

It is so weird because sometime the tests works well but 

We use protractor and custom framework is cucumber. We run our suites paralell at the same time. 

I have downgraded some libriaries but it have not resolved my problem.

Here is my getting error;

```
Error: function timed out, ensure the promise resolves within 180000 milliseconds
    at Timeout._onTimeout (/var/lib/jenkins/workspace/E2E/Docker-Selenium/Run-Selenium-Suites/e2e/node_modules/cucumber/src/user_code_runner.js:61:18)
    at listOnTimeout (timers.js:327:15)
    at processTimers (timers.js:271:5)
```

Bug report
---

- Node Version: 11.10.0
- Protractor Version: 5.4.2
- Angular Version: -
- Browser(s): chrome
- Operating System and Version Linux Ubuntu 16
- Your protractor configuration file
- A relevant example test
- Output from running the test
- Steps to reproduce the bug
- The URL you are running your tests against (if relevant)

Here is my package.json file ;

```
 "dependencies": {
    "archiver": "^2.1.1",
    "nodemailer": "^4.7.0",
    "protractor-cucumber-framework": "^6.1.1",
    "sprintf-js": "^1.1.2"
  },
  "devDependencies": {
    "babel-plugin-transform-runtime": "^6.23.0",
    "chai": "^4.2.0",
    "chai-as-promised": "^7.1.1",
    "chrome-remote-interface": "^0.27.0",
    "cucumber": "^4.2.1",
    "cucumber-junit": "^1.7.1",
    "cucumber-pretty": "^1.5.0",
    "exceljs": "^1.6.3",
    "exceljs-plus": "0.0.5",
    "fs-extra": "^7.0.1",
    "got": "^9.5.0",
    "grunt": "^1.0.3",
    "grunt-cli": "^1.3.2",
    "grunt-protractor-runner": "^5.0.0",
    "gulp": "^3.9.1",
    "moment-business-days": "^1.1.3",
    "pdfjs-dist": "^2.0.943",
    "protractor": "^5.4.1",
    "protractor-flake": "^3.3.0",
    "protractor-multiple-cucumber-html-reporter-plugin": "^1.8.0",
    "readline-specific": "^1.0.4",
    "readline-promise": "^1.0.3",
    "soap": "^0.25.0",
    "uncaught": "0.0.5",
    "underscore": "^1.9.1",
    "user-info": "^1.0.0",
    "webdriver-manager": "^12.1.0"
  }
```

Last getting error method block is;

```
When(/^İhbar kaydı dogrulama altında Yeni Olay Tanımlama seçilir$/, async function () {
    await PageObject.newEvent.click();

    return await WorldPage.aydesAjaxLoading.waitForInvisibilityOf();
});
```

Here is my Extenders.js spesific method:

```
protractor.ElementFinder.prototype.waitForInvisibilityOf = async function (timeout) {
    const self = this;
    let _debug = false;
    let isDone = false;
    timeout = timeout || WAITTIMEOUT;
    let seconds = timeout / 1000;
    let style, invisibilityOf, visibilityOf;

    if (await !self.isPresent())
        return;

    //This loop will rotate for 60 times to check If page Is ready after every 1 second.
    //You can replace your value with 60 If you wants to Increase or decrease wait time.
    for (let i = 1; i <= seconds; i++) {
        await browser.sleep(1000);

        style = await self.getAttribute('style');
        invisibilityOf = await style.includes('display: none;');
        visibilityOf = await style.includes('display: block;');

        if (invisibilityOf) {
            if (_debug)
                console.log(i + " second: Element invisible!");
            isDone = true;
        }
        else {
            if (_debug)
                console.log(i + " second: Element NOT invisible!");
        }

        if (seconds === i)
            throw ", Element invisibility timed out after " + timeout + " milliseconds";

        if (!invisibilityOf && !visibilityOf && i > 5) // break for paging is loaded
            break;

        if (isDone) // If element is invisible
            break;
    }
};
```

Thank you.

**Comment by GitSage** [comment]
This error is coming [from CucumberJS](https://github.com/cucumber/cucumber-js/blob/509e64fd5d9835c9101b33a642e6ccac459025b8/src/user_code_runner.js). Any particular reason you suspect Protractor to be the problem?

Cucumber throws that error when a single step takes longer than its configured timeout to complete. In your case you've configured it to 180000ms = 3 minutes. I don't immediately see any reason why your script would take that long, but presumably it is doing so. 

I suggest switching to Protractor's built-in [invisibilityOf](https://www.protractortest.org/#/api?view=ProtractorExpectedConditions) implementation. Why make your own implementation when there's already an officially supported one?

**Comment by espekkaya** [comment]
Hello GitSage,

Firstly, I suspect protractor because when I added every line to console.log in waitForInvisibilityOf  method. The weird part of it that most case I got an error below line. And also I got promise error on protractor click method as well.

`style = await self.getAttribute('style');`

In our project structure it does not solve wait loading panel using invisibilityof method so we have to use our workaround solve this problem.

Yesterday I have downgraded nodejs version 11.10 to 10.15. In that case it's working good for now. Little code getting promise resolve error but it has not blocked other test suite.
